OSASINFRA-3675: Rework authentication in Manila CSI Driver Operator

dkokkino opened the pull request:
- Updated manila assets to mount secret and configmap to node/controller
- Modified the secret generator to use different parameters in preparation for change to manila driver

Commit:
Update manila assets for new authentication method

- Adding clouds.yaml as a secret mount volume. This will be used for authentication by the manila driver to match how cinder does authentication.

Commit:
Modify secret generator

 - Goal is to change how the aunthentication for the manila driver
 occurs.Changes the fields generated in the secret csi-manila-secrets

stephenfin commented on pkg/openstack-manila/secret/secretsync.go:
```diff
@@ -137,54 +138,14 @@ func (c *SecretSyncController) translateSecret(cloudSecret *v1.Secret) (*v1.Secr
+	data["os-use-clouds"] = []byte(strconv.FormatBool(cloud.UseClouds))
+
+	if cloud.CloudsFile != "" {
+		data["os-clouds-file"] = []byte(cloud.CloudsFile)
```
This indicates that path that the `clouds.yaml` can be found in. We're using this value in the Manila CSI driver, thus the value of this should be static and correspond to the path you are mounting the `clouds.yaml` file to.

Looking at your manifests from [an earlier commit](https://github.com/openshift/csi-operator/pull/373/commits/94000a8679fc362167ae1ee62da618a1c913d9c3#diff-f48ca1ef12471e0a934482bc5478ba4174bedef8677429c13815be4abed8ef7fR63), that file is currently `/etc/kubernetes/secret/clouds.yaml` (though I'm suggesting to change it). You should use the same static path here.

stephenfin commented on assets/overlays/openstack-manila/patches/node_add_driver.yaml:
```diff
@@ -57,6 +57,12 @@ spec:
             - name: MANILA_SHARE_PROTO
               value: NFS
           volumeMounts:
+            - name: config-manilaplugin
+              mountPath: /etc/kubernetes/config
+              readOnly: true
```
Do you want this? We don't have support for a `cloud.conf` file in the Manila driver yet, right?

stephenfin commented on assets/overlays/openstack-manila/patches/node_add_driver.yaml:
```diff
@@ -86,6 +92,18 @@ spec:
               memory: 50Mi
           terminationMessagePolicy: FallbackToLogsOnError
       volumes:
+        - name: config-manilaplugin
```
Per above, I'm not sure you want this.

stephenfin commented on assets/overlays/openstack-manila/patches/node_add_driver.yaml:
```diff
@@ -86,6 +92,18 @@ spec:
+            items:
+              - key: cloud.conf
+                path: cloud.conf
+        - name: secret-manilaplugin
```
I know you're aligning with Cinder again here, but would a more descriptive name be helpful? Say, `cloud-credentials`? Again, I'm planning on doing the same thing for Cinder.

stephenfin commented on assets/overlays/openstack-manila/patches/node_add_driver.yaml:
```diff
@@ -57,6 +57,12 @@ spec:
+              mountPath: /etc/kubernetes/config
+              readOnly: true
+            - name: secret-manilaplugin
+              mountPath: /etc/kubernetes/secret
```
Purely a nice to have, but `clouds.yaml` are typically found in `/etc/openstack`. I know you're aligning with Cinder here - which is good - but I wonder if we could use that here?

Note: I am planning on changing this in Cinder too

stephenfin reviewed (changes requested):
Looks like you need to run `make verify` too

Commit:
Create static secret to replace sycretsync

 - Previously the csi-manila-secrets secret in the
 openshift-manila-csi-driver namespace required secretsync.go to
 generate the secret dynamically. Since we are changing how the
 authentication occurs for the manila driver that process can be
 replaced with a static secret asset.

pierreprinetti commented:
/retest-required

dkokkino commented:
/retest-required

dkokkino commented:
/retest

dkokkino commented:
/retest

dkokkino commented:
/retest

MiguelCarpio commented:
/test e2e-openstack

dkokkino commented:
/test e2e-openstack-manila-csi

MiguelCarpio commented:
/test e2e-openstack-manila-csi

stephenfin commented:
/approve
/lgtm
/hold

I've tested this locally with https://github.com/openshift/hypershift/pull/6683 and HCP deploys just fine. @dkokkino please remove the hold once https://github.com/openshift/hypershift/pull/6683 has merged

dkokkino commented:
/test hypershift-e2e-openstack-aws-csi-manila

dkokkino commented:
/test hypershift-e2e-openstack-aws-csi-manila

stephenfin commented:
I am still investigating the hypershift failures. Just leaving notes here for myself. Currently this is failing with:

```
E0822 13:41:55.805285       1 driver.go:331] [ID:594] GRPC error: rpc error: code = Unauthenticated desc = failed to create Manila v2 client: failed to authenticate: Get "https://38.102.83.157:13000/": tls: failed to verify certificate: x509: certificate signed by unknown authority
```

From the `csi-driver` container of the `clusters-dfc4bbcb3230eac907a2 / openstack-manila-csi-controllerplugin-57596d7957-kjkh` pod. That pod defines its volumes like so:

```
apiVersion: v1
kind: Pod
...
spec:
  ...
  volumes:
    ...
    - name: cloud-credentials
      secret:
        defaultMode: 420
        items:
          - key: clouds.yaml
            path: clouds.yaml
        secretName: manila-cloud-credentials
    - configMap:
        defaultMode: 420
        items:
          - key: ca-bundle.pem
            path: ca-bundle.pem
        name: openstack-cloud-config
        optional: true
      name: cacert
```

I am attempting to fix this in https://github.com/openshift/hypershift/pull/6777 but the approach I've taken is wrong (per that LLM-generated warning) so I'll need to take another punt at it.

stephenfin commented:
Weird. I deployed this locally and things appear to be working as-is. My pod definitions are the same:

```yaml
apiVersion: v1
kind: Pod
metadata:
  # ...
  name: openstack-manila-csi-controllerplugin-5994b65bf7-q2x8x
  namespace: clusters-stephenfin-hcp
  # ...
spec:
  # ...
  volumes:
    # ...
    - name: cloud-credentials
      secret:
        defaultMode: 420
        items:
          - key: clouds.yaml
            path: clouds.yaml
        secretName: manila-cloud-credentials
    - configMap:
        defaultMode: 420
        items:
          - key: ca-bundle.pem
            path: ca-bundle.pem
        name: openstack-cloud-config
        optional: true
    # ...
```

I see the credentials in the correct place with the correct key:

```yaml
❯ oc get -n clusters-stephenfin-hcp secret manila-cloud-credentials -o yaml
apiVersion: v1
data:
  cloud.conf: <redacted>
  clouds.yaml: <redacted>
kind: Secret
metadata:
  annotations:
    hypershift.openshift.io/cluster: clusters/stephenfin-hcp
  creationTimestamp: "2025-09-16T11:00:30Z"
  name: manila-cloud-credentials
  namespace: clusters-stephenfin-hcp
  resourceVersion: "2080159"
  uid: 5ba57428-8af4-4b04-9dae-acb9f55c2bf2
type: Opaque
```

I don't see a `ca-bundle.pem` key in the config map, but that's expected since the cloud I'm on isn't using self-signed certs. And it *was* in the correct place in the 

```yaml
❯ oc get -n clusters-stephenfin-hcp cm openstack-cloud-config -o yaml
apiVersion: v1
data:
  cloud.conf: |
    <redacted>
kind: ConfigMap
metadata:
  creationTimestamp: "2025-09-16T11:04:40Z"
  name: openstack-cloud-config
  namespace: clusters-stephenfin-hcp
  ownerReferences:
  - apiVersion: hypershift.openshift.io/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: HostedControlPlane
    name: stephenfin-hcp
    uid: 1ab35480-3265-44e9-aed2-d890112f3018
  resourceVersion: "2081779"
  uid: 72ab196c-85af-49c0-b58a-560d9ee54752
```

So I'm not really sure what's going on currently 😕 I wonder if the CA cert is old?

dkokkino commented:
/retest e2e-openstack

eshulman2 commented:
/retest

stephenfin commented:
/jira refresh

stephenfin commented:
/jira refresh

stephenfin commented:
/jira refresh

stephenfin commented:
/verified later

Hypershift CI is perma-failing at the moment, so there isn't much we can do about this. We'll need to manually test that integration when we get there

stephenfin commented:
/verified later @stephenfin

Hypershift CI is perma-failing at the moment, so there isn't much we can do about this. We'll need to manually test that integration when we get there

stephenfin commented:
/lgtm

stephenfin commented:
/remove-hold

stephenfin commented:
/test e2e-openstack

This is just a slow disk.

/override ci/prow/hypershift-e2e-openstack-aws-csi-manila

This is known failure due to CAPI API version mismatches (which we need the HCP team to fix)

stephenfin commented:
/test e2e-openstack

We need to get off of this cloud ASAP.

stephenfin commented:
/test ci-index-aws-efs-csi-driver-operator-bundle

This is unrelated
